Simplification of the loading process for rasterized tiles and mesh tiles

gchoqueux opened the pull request:
## Description

This PR proposes a possible simplification of the loading process for **rasterized tiles** and **mesh tiles**.  

### Summary of changes

- `TileMesh` instances are now created **directly in** `TiledGeometryLayer`, **without using commands or providers**.  
- `TileMesh.pendingSubdivision` has been **removed**.  
- `TiledGeometryLayer#hasEnoughTexturesToSubdivide` has been **removed**.  
- `updateLayeredMaterialNodeImagery` and `updateLayeredMaterialNodeElevation` have been **merged into a single lightweight method**.  
- `TileProvider` has been **deleted**.  
- Texture loading is now handled by **`RasterTile#load`**.

fix #2816

gchoqueux commented:
Rebased

jailln commented:
> @jailln I solved my problem, which was mainly caused by excessive data loading. The functional tests work with this PR, there isn't issue like #2530

Oh ok I think I know why: in #2521 tilemesh creation doesn't go through the scheduler anymore (since there is no external query, it is not required), that's why it breaks the events. We saw a perf improvement with this change so it needs to be considered in my opinion.

Desplandis commented on packages/Main/src/Layer/ElevationLayer.js:
```diff
@@ -115,15 +114,16 @@ class ElevationLayer extends RasterLayer {
      */
     setupRasterNode(node) {
-        const rasterElevationNode = new RasterElevationTile(this);
+        const tiles = node.getExtentsByProjection(this.crs);
```
TODO: assert this code.

Desplandis commented on packages/Main/src/Renderer/RasterTile.js:
```diff
@@ -1,9 +1,43 @@
+
+    // to do remove useless
+    return !cmd.force && (!view || !requester.parent || !requester.material || !requester.material.visible || child?.visible);
+}
```
The previous conditions for aborting a command were the following. I marked those that are still here in this PR.
- [x] `TileMesh` parent is not defined (which surely means it is marked for deletion) or the material is not visible (which means that we have a more precise LOD here). Both conditions are here.
- [ ] The `TileMesh` has a more precise elevation texture attached, so we cancel this request (I imagine this is the case of SRTM vs USGS layer). We don't have this condition anymore, any reasons?
- [ ] The layer was removed, which was given by two fragile conditions: no update state on the node for this layer or a more fragile condition the _featuresCaches does not exist anymore in the source... I agree that we should remove both of those conditions but do we still have a way to know that the layer was de-attached/disposed?

Desplandis commented on packages/Main/src/Layer/RasterLayer.js:
```diff
@@ -44,6 +61,23 @@ class RasterLayer extends Layer {
+
+    update(context, layer, node) {
+        if (layer.visible && !layer.freeze && this.anyVisibleData(node)) {
+            const rasterTile = node.material.getTile(this.id) || this.setupRasterNode(node);
```
So `setupRasterNode` is an abstract method of `RasterLayer`

Desplandis commented on packages/Main/src/Layer/ElevationLayer.js:
```diff
@@ -135,11 +135,16 @@ class ElevationLayer extends RasterLayer {
+    anyVisibleData(node) {
+        const current = node.material.getElevationTile()?.layer;
+        return super.anyVisibleData(node)
+            && (current?.source == this.source || this.source.zoom.min > (current?.source.zoom.max || -1));
```
What is the meaning of `-1` here?

gchoqueux committed:
fix(test): issues from refactoring and rebasing

gchoqueux committed:
fix(test): issues from refactoring and rebasing

gchoqueux committed:
refacto(RasterTile): simplify raster layer loading

BREACKING CHANGES: remove pdateLayeredMaterialNodeImagery, updateLayeredMaterialNodeElevation

gchoqueux committed:
refacto(examples): use await top level to fetch

gchoqueux committed:
refacto(RasterTile): simplify raster layer loading

BREACKING CHANGES: remove pdateLayeredMaterialNodeImagery, updateLayeredMaterialNodeElevation
